File: classes04/cpp/blinkDetect.cpp
#include <opencv2/objdetect.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/plot.hpp>
#include <opencv2/opencv.hpp>

#include <dlib/opencv.h>
#include <dlib/image_processing/frontal_face_detector.h>
#include <dlib/image_processing.h>
#include <dlib/image_io.h>
#include <dlib/gui_widgets.h>


using namespace std;
using namespace cv;
using namespace dlib;

// 全局参数
int state =0;
double thresh = 0.43; // 判断闭眼的阈值
float normallizeCount =0;
int eyeRegionCount = 0;
string modelPath = "../../common/resources/zxm_shape_predictor_70_face_landmarks.dat"; //形状检测模型
shape_predictor poseModel; //形状检测器
frontal_face_detector detector=dlib::get_frontal_face_detector(); //脸部检测器

Mat frame, eye;
#define RESIZE_HEIGHT 480;
#define FACE_DOWNSAMPLE_RATIO_DLIB 2.5

float blinkTime = 0.2;  // 眨眼的事件限制是200ms
float drowsyTime = 1.0; // 疲劳的时间限制是1000ms
int drowsyLimit = 0;
int falseBlinkLimit = 0;

static int lefteye[]={36,37,38,39,40,41};
std::vector<int> lefteye_index(lefteye, lefteye + sizeof(lefteye)/sizeof(lefteye[0]));
static int righteye[]={42,43,44,45,46,47};
std::vector<int> righteye_index(righteye, righteye + sizeof(righteye)/sizeof(righteye[0]));

/**
 * @brief 检查眼睛的状态是睁开的还是闭起来的
 * 
 * @param landmarks 关键点
 * @return int 0:闭，1:开
 */
int checkEyeStatus(full_object_detection& landmarks)
{
    // 创建一个黑色的图片作为眼睛的遮罩
    Mat mask = Mat::zeros(frame.rows, frame.cols, frame.depth());
    std::vector<Point> hullLeftEye;
    for(int i =0; i < lefteye_index.size(); i++)
    {
        Point pt(landmarks.part(lefteye_index[i]).x(), landmarks.part(lefteye_index[i]).y());
        hullLeftEye.push_back(pt);
    }
    fillConvexPoly(mask, hullLeftEye, cv::Scalar(255, 255, 255));

    std::vector<Point> hullRightEye;
    for(int i =0; i < righteye_index.size(); i++)
    {
        Point pt(landmarks.part(righteye_index[i]).x(), landmarks.part(righteye_index[i]).y());
        hullRightEye.push_back(pt);
    }
    fillConvexPoly(mask, hullRightEye, cv::Scalar(255, 255, 255));

    // 眼睛的长度
    int lenLeftEyeX = landmarks.part(lefteye_index[3]).x() - landmarks.part(lefteye_index[0]).x();
    int lenLeftEyeY = landmarks.part(lefteye_index[3]).y() - landmarks.part(lefteye_index[0]).y();
    float lenLeftEyeSquart = lenLeftEyeX*lenLeftEyeX + lenLeftEyeY*lenLeftEyeY;

    eyeRegionCount = cv::countNonZero(mask == 255);
    
    // 使用眼睛的长度对数值进行归一化
    normallizeCount = (float)eyeRegionCount/lenLeftEyeSquart;

    eye = Mat::zeros(frame.rows, frame.cols, frame.depth());
    frame.copyTo(eye, mask);

    int eyeStatus = 1;

    if(normallizeCount < thresh)
        eyeStatus = 0;

    return eyeStatus;
}

/**
 * @brief 使用FSM追踪用户的眨眼状态，推算用户的行为
 * 
 * @param eyeStatus 眼睛的状态
 * @param blinkCount 眨眼的次数
 * @param drowsy 疲劳
 * @return int 
 */
int checkBlinkStatus(int eyeStatus, int& blinkCount, int& drowsy)
{
    if(state >=0 && state <= falseBlinkLimit)
    {
        // 如果在当前帧眼睛是睁开的，那就继续保持;判断是否是无效的眨眼
        if(eyeStatus)
        {
            state = 0;
        }
        else
        {
            state++;
        }
    }
    else if(state > falseBlinkLimit && state <= drowsyLimit)
    {
        if(eyeStatus)
        {
            state = 0;
            blinkCount++;
            return 1;
        }
        else
        {
            state++;
        }
    }
    else
    {
        if(eyeStatus)
        {
            state = 0;
            blinkCount++;
            drowsy = 0;
            return 1;
        }
        else
        {
            drowsy =1;
        }
    }
    return 0;
}

int main(int argc, char const *argv[])
{
    cout<<"注意: ./blinkDetect <默认的阈值为："<< thresh <<endl;
    
    if(argc == 2) //自定义阈值
    {
        thresh = atof(argv[1]);
    }

    VideoCapture capture;
    deserialize(modelPath) >> poseModel;
    std::vector<dlib::rectangle> faces;

    capture.open(0);

    // 计算帧率
    int blinkCount = 0;
    int drowsy =0;
    double t=0;

    // 首先空跑几帧等待摄像头状态稳定
    for(int i=0; i<50; i++)
        capture.read(frame);
    
    float totalTime = 0.0;
    int validFrames = 0;
    int dummyFrames = 50;
    float spf = 0;
    double timeLandmarks = 0.0;

    while(validFrames < dummyFrames)
    {
        t = (double)cv::getTickCount();

        capture.read(frame);
        validFrames++;
        int height = frame.rows;
        float IMAGE_SIZE =(float)height/RESIZE_HEIGHT;

        cv::resize(frame, frame, cv::Size(), 1.0/IMAGE_SIZE, 1.0/IMAGE_SIZE);
        cv::Size size = frame.size();
        cv::Mat frame_small;
        cv::resize(frame, frame_small, cv::Size(),1.0/FACE_DOWNSAMPLE_RATIO_DLIB, 1.0/FACE_DOWNSAMPLE_RATIO_DLIB);

        cv_image<bgr_pixel> cimg(frame);
        cv_image<bgr_pixel> cimg_small(frame_small);

        // 检测脸部
        faces = detector(cimg_small);

        timeLandmarks = ((double)cv::getTickCount() - t)/cv::getTickFrequency();

        // 检查帧是否有效
        if(!faces.size())
        {
            validFrames--;
            putText(frame,"不能检测到面部，请检查!", cv::Point(10,50),FONT_HERSHEY_SIMPLEX,0.4,cv::Scalar(0,0,255));
            putText(frame,"或者采样的比例设置的太小了!", cv::Point(10,50),FONT_HERSHEY_SIMPLEX,0.4,cv::Scalar(0,0,255));
            imshow("Blink Detection Demo", frame);
            if((waitKey(1)&0xFF) == 27)
            {
                return 0;
            }
        }
        else
            totalTime += timeLandmarks;
    }

    spf = totalTime/dummyFrames;
    cout<<"SPF(每帧用时):"<<spf<<"秒"<<endl;

    /**
     * 使用spf计算眨眼的帧数显示和疲劳的帧数限制，这些参会用于FSM状态检测
     */
    drowsyLimit =int(drowsyTime/spf);
    falseBlinkLimit = int(blinkTime/spf);
    cout<<"疲劳帧数限制:"<<drowsyLimit<<"帧("<<drowsyLimit*spf*1000<<" ms)"<<endl;
    cout<<"假眨眼的帧数限制:"<<falseBlinkLimit<<"帧("<<falseBlinkLimit*spf*1000<<" ms)"<<endl;

    if(!capture.isOpened())
    {
        cout<<"错误：摄像头打开失败!"<<endl;
        return -1;
    }

    while(capture.read(frame))
    {
        if(frame.empty())
        {
            cout<<"警告:未获取到有效的像素"<<endl;
            break;
        }

        int height = frame.rows;
        float IMAGE_RESIZE = (float)height/RESIZE_HEIGHT;
        cv::resize(frame, frame, cv::Size(), 1.0/IMAGE_RESIZE, 1.0/IMAGE_RESIZE);
        cv::Size size = frame.size();

        cv::Mat frame_small;
        cv::resize(frame, frame_small, cv::Size(), 1.0/FACE_DOWNSAMPLE_RATIO_DLIB, 1.0/FACE_DOWNSAMPLE_RATIO_DLIB);
        cv_image<bgr_pixel> cimg(frame);
        cv_image<bgr_pixel> cimg_small(frame_small);

        faces = detector(cimg_small);
        if(!faces.size())
        {
            putText(frame,"不能检测到面部，请检查!", cv::Point(10,50),FONT_HERSHEY_SIMPLEX,0.4,cv::Scalar(0,0,255));
            putText(frame,"或者采样的比例设置的太小了!", cv::Point(10,50),FONT_HERSHEY_SIMPLEX,0.4,cv::Scalar(0,0,255));
            imshow("Blink Detection Demo", frame);
            if((waitKey(1)&0xFF) == 27)
            {
                return 0;
            }
            continue;
        }
        dlib::rectangle r(
                            (long)(faces[0].left()*FACE_DOWNSAMPLE_RATIO_DLIB),
                            (long)(faces[0].top()*FACE_DOWNSAMPLE_RATIO_DLIB),
                            (long)(faces[0].right()*FACE_DOWNSAMPLE_RATIO_DLIB),
                            (long)(faces[0].bottom()*FACE_DOWNSAMPLE_RATIO_DLIB)
                         );

        full_object_detection landmarks;
        landmarks = poseModel(cimg, r);

        // 检查用户的眼睛是睁开的还是闭起来的
        int eyeStatus = checkEyeStatus(landmarks);

        // 将眼睛的状态传入FSM来确定眨眼的状态
        int blinkStatus = checkBlinkStatus(eyeStatus, blinkCount, drowsy);

        for(int i=0;i<righteye_index.size();i++)
        {
            cv::Point pt(landmarks.part(righteye_index[i]).x(), landmarks.part(righteye_index[i]).y() );
            circle(frame, pt, 1, Scalar(255, 0, 255), 1, 8);
        }
        for(int i=0;i<lefteye_index.size();i++)
        {
            cv::Point pt(landmarks.part(lefteye_index[i]).x(), landmarks.part(lefteye_index[i]).y() );
            circle(frame, pt, 1, Scalar(255, 0, 255), 1, 8);
        }

        if(drowsy)
        {
            putText(eye, cv::format("state:%d , blinks:%d", state, blinkCount), cv::Point(10, 50), cv::FONT_HERSHEY_COMPLEX, 0.9, cv::Scalar(0, 255, 255), 1);
            putText(frame, cv::format("!!! DROWSY !!!"), cv::Point(50,100), cv::FONT_HERSHEY_COMPLEX,1.2, cv::Scalar(0, 0, 255),1);
        }
        else
        {
            putText(eye, cv::format("state:%d , blinks:%d", state, blinkCount), cv::Point(10, 50), cv::FONT_HERSHEY_COMPLEX, 0.9, cv::Scalar(0, 255, 255), 1);
            putText(frame, cv::format("blinks: %d", blinkCount), cv::Point(50,100), cv::FONT_HERSHEY_COMPLEX,0.9, cv::Scalar(0, 255, 255),1);
        }

        imshow("Blink Detection Demo", frame);
        imshow("Eye", eye);
        char c= (char)waitKey(1);
        if(c==27)
        {
            break;
        }

    }

    capture.release();
    destroyAllWindows();


    return 0;
}

File: classes04/python/blinkDetect.py
import cv2
import dlib
import time
import sys
import numpy as np

FACE_DOWNSAMPLE_RATIO = 1.5
RESIZE_HEIGHT = 360

thresh = 0.43

# 全局变量
modelPath = "../../common/resources/zxm_shape_predictor_70_face_landmarks.dat"
detector = dlib.get_frontal_face_detector()
predictor = dlib.shape_predictor(modelPath)

# 眼睛的dlib索引点
leftEyeIndex = [36, 37, 38, 39, 40, 41]
rightEyeIndex = [42, 43, 44, 45, 46, 47]

# 计算FPS的变量
blinkCount = 0
drowsy = 0
state =0
blinkTime = 0.2  #200ms
drowsyTime = 1.0 #1000ms

# 检查眼睛的状态
def checkEyeStatus(landmarks):
    """
    检查眼睛是睁开还是闭上
    参数:
    =============
    landmarks:形状的关键特征点
    返回:
    =============
    1->眼镜睁开；0->眼睛闭上;
    """
    pass

    # 创建一个黑色的遮罩用来截取眼镜区域
    mask = np.zeros(frame.shape[:2], dtype=np.float32)

    hullLeftEye=[]
    for i in range(0, len(leftEyeIndex)):
        hullLeftEye.append((landmarks[leftEyeIndex[i]][0], landmarks[leftEyeIndex[i]][1]))
    cv2.fillConvexPoly(mask, np.int32(hullLeftEye), 255)

    hullRightEye = []
    for i in range(0, len(rightEyeIndex)):
        hullRightEye.append((landmarks[rightEyeIndex[i]][0], landmarks[rightEyeIndex[i]][1]))
    cv2.fillConvexPoly(mask, np.int32(hullRightEye), 255)

    lenLeftEyeX = landmarks[leftEyeIndex[3]][0] - landmarks[leftEyeIndex[0]][0]
    lenLeftEyeY = landmarks[leftEyeIndex[3]][1] - landmarks[leftEyeIndex[0]][1]

    lenLeftEyeSquare = lenLeftEyeX*lenLeftEyeX + lenLeftEyeY*lenLeftEyeY

    eyeRegionCount = cv2.countNonZero(mask)

    normalizedCount = eyeRegionCount/np.float32(lenLeftEyeSquare)

    eyeStatus = 1
    if(normalizedCount < thresh):
        eyeStatus = 0

    return eyeStatus

def checkBlinkStatus(eyeStatus):
    global state, blinkCount, drowsy

    # 睁开状态和假眨眼状态
    if(state >=0 and state <= falseBlinkLimit):
        # 如果眼睛是睁开的
        if(eyeStatus):
            state=0
        # 如果还是闭着的
        else:
            state +=1
    elif(state>falseBlinkLimit and state <= drowsyLimit):
        if(eyeStatus):
            state = 0
            blinkCount += 1
        else:
            state +=1
    else:
        if(eyeStatus):
            state = 0
            blinkCount +=1
            drowsy =0
        else:
            drowsy =1

def getLandmarks(im):
    imSmall = cv2.resize(im, None, fx=1.0/FACE_DOWNSAMPLE_RATIO, fy=1.0/FACE_DOWNSAMPLE_RATIO, interpolation=cv2.INTER_LINEAR)

    rects = detector(imSmall, 0)
    if len(rects) == 0:
        return 1
    newRect = dlib.rectangle(int(rects[0].left()*FACE_DOWNSAMPLE_RATIO),
                             int(rects[0].top()*FACE_DOWNSAMPLE_RATIO),
                             int(rects[0].right()*FACE_DOWNSAMPLE_RATIO),
                             int(rects[0].bottom()*FACE_DOWNSAMPLE_RATIO))

    points=[]
    [points.append((p.x, p.y)) for p in predictor(im, newRect).parts()]

    return points

capture = cv2.VideoCapture(0)

for i in range(50):
    ret, frame = capture.read()

totalTime = 0.0
validFrames = 0
dummyFrames = 50
spf = 0

while(validFrames < dummyFrames):
    validFrames +=1
    t = time.time()
    ret, frame = capture.read()
    height, width = frame.shape[:2]
    IMAGE_RESIZE = np.float32(height)/RESIZE_HEIGHT
    frame = cv2.resize(frame, None, fx=1.0/IMAGE_RESIZE, fy=1.0/IMAGE_RESIZE, interpolation=cv2.INTER_LINEAR)

    landmarks = getLandmarks(frame)
    timeLandmarks = time.time() -t

    # 未检测到人
    if landmarks == 1:
        validFrames -=1
        cv2.putText(frame, "Unable to detect face, Please check proper lighting", (10, 50), cv2.FONT_HERSHEY_COMPLEX, 0.9, (0, 255, 255), 1, cv2.LINE_AA)
        cv2.putText(frame, "Or Decrease FACE_DOWNSAMPLE_RATIO", (10, 50), cv2.FONT_HERSHEY_COMPLEX, 0.9, (0, 255, 255), 1, cv2.LINE_AA)
        cv2.imshow("Blink Detection Demo", frame)

        if cv2.waitKey(1)&0xFF == 27:
            sys.exit()
    else:
        totalTime += timeLandmarks
        
spf = totalTime/dummyFrames
print("SPF为:{:.2f} ms".format(spf*1000))

drowsyLimit = drowsyTime/spf
falseBlinkLimit = blinkTime/spf
print("drowsyLimit:{}, falseBlinkLimit:{}".format(drowsyLimit, falseBlinkLimit))

while(1):
    try:
        t=time.time()
        ret, frame = capture.read()
        height, width = frame.shape[:2]
        IMAGE_RESIZE = np.float32(height)/RESIZE_HEIGHT
        frame = cv2.resize(frame, None, fx=1.0/IMAGE_RESIZE, fy=1.0/IMAGE_RESIZE, interpolation=cv2.INTER_LINEAR)
        landmarks= getLandmarks(frame)
        if landmarks==1:
            cv2.putText(frame, "Unable to detect face, Please check proper lighting", (10, 50), cv2.FONT_HERSHEY_COMPLEX, 0.9, (0, 255, 255), 1, cv2.LINE_AA)
            cv2.putText(frame, "Or Decrease FACE_DOWNSAMPLE_RATIO", (10, 50), cv2.FONT_HERSHEY_COMPLEX, 0.9, (0, 255, 255), 1, cv2.LINE_AA)
            cv2.imshow("Blink Detection Demo", frame)

            if cv2.waitKey(1)&0xFF == 27:
                sys.exit()
            continue
        
        eyeStatus = checkEyeStatus(landmarks)
        checkBlinkStatus(eyeStatus)

        for i in range(0, len(leftEyeIndex)):
            cv2.circle(frame, (landmarks[leftEyeIndex[i]][0], landmarks[leftEyeIndex[i]][1]), 1, (255,0,255), thickness=1, lineType=cv2.LINE_AA)
        for i in range(0, len(rightEyeIndex)):
            cv2.circle(frame, (landmarks[rightEyeIndex[i]][0], landmarks[rightEyeIndex[i]][1]), 1, (255,0,255), thickness=1, lineType=cv2.LINE_AA)
        
        if(drowsy):
            cv2.putText(frame, "!!! Drowsy !!!", (50, 100), cv2.FONT_HERSHEY_COMPLEX, 1.0, (0, 0, 255), thickness=1, lineType=cv2.LINE_AA)
        else:
            cv2.putText(frame, "Blink:{}".format(blinkCount), (50, 50), cv2.FONT_HERSHEY_COMPLEX, 1.0, (0, 255, 255), thickness=1, lineType=cv2.LINE_AA)
            pass
        
        cv2.imshow("Blink Detection Demo", frame)
        
        if cv2.waitKey(1)&0xFF == 27:
            break
        print("耗时:", time.time() - t)
    
    except Exception as e:
        print(e)

capture.release()
cv2.destroyAllWindows()

    
